Adds count viewport setting

nkakouros opened the pull request:
This adds a viewport option to limit the number of tasks shown under the viewport:

```
# Show only 4 tasks | project:my-project *4
```

My use case is that I want to have a "daily tasks" file that gets updated with the top 2 most urgent tasks across all projects. I tried to experiment with the `limit:4` taskwarrior filter but this gets applied during `export` and does not take viewport or any other sorting into account.

I tried to add tests but I am not familiar with writing tests and I have no idea if it is correct.

nkakouros commented:
@tbabej @liskin This PR is ready to be merged if you find it useful.

liskin commented:
I don't think I will find use for it personally but it seems simple enough that I'm willing to help reviewing it. To do so, I'd appreciate it if you rebase this on top of current master (instead of merging master) and get rid of unrelated changes (there are some unnecessary whitespace changes and quote changes).

tbabej commented:
Hey @nkakouros, this looks like a useful addition and I would be happy to include it in the upstream. However, I am a bit cautious about introducing new syntactic elements to the header definition - over time, the header syntax could just become too complicated for its own good.

Instead of introducing new syntax element (namely `*<limit>` to the header), what about just "overriding" the behavior of the `limit:<limit>` clause that TW users are already familiar with. In short:

1. Detect if there is a `limit:...` clause in the filter query
2. Do not propagate it down to TW, but just keep a record about the limit in the viewport
3. Apply the limit taking the taskwiki settings (such as sort order) into account

nkakouros commented:
In this case, we assume that no one will want to use the `limit` with the default `export` command, right? I guess this is valid assumption. But how would this make the header simpler? The python code would still do exactly the same things, just the users would use `limit` instead of `*`. This is more keeping the cognitive load for the users a tiny bit less, than keeping things simple for taskwiki. 

Is the above your thought also?

nkakouros commented:
Now that we are on it, I am planning the following PRs also (some have already been opened, I will add the tests):

- supporting custom priority values in taskwarrior (haven't settled exactly on how it will work with `!`). This is useful for users like me that have added for instance one more priority in taskwarrior and which causes errors currently in taskwiki.
- a new `expires` element in the header that defines the last date that taskwiki should mess with a viewport. The syntax is `!2020-01-01`, an exclamation mark with a date. This is useful for creating records of e.g. meeting notes and in general for archiving. You can have a tasks tagged with e.g. `+meeting`, you have them show up in a doc with a viewport like `# Meeting agenda | +meeting  created:2-weeks-ago-or-sth !2020-meeting-date` and be sure that when you reopen this to see what you discusses, what was done, what new tasks were decided, you will not see a current view, but the archived one. 


Do you think the above will add overhead, are not useful or want a different syntax?

nkakouros commented:
@liskin I addressed your comments.

@tbabej The last commit uses `limit:` instead of `*`. I prefer the looks of the old syntax, but you run the show.

tbabej commented:
@nkakouros Regarding the other PRs, I am fine with those proposals on the high level. Maybe some concerns will show up during the review, but overall I think these are useful ideas.

Thanks for converting this PR to use `limit:`.

tbabej commented on taskwiki/regexp.py:
```diff
@@ -44,47 +44,47 @@
-        r'\s*'                       # Any whitespace
-        r'(\$(?P<sort>[A-Z]))?'      # Optional sort indicator
-        r'\s*'                       # Any whitespace
-        r'(\*(?P<count>[0-9]+))?'    # Optional count indicator
```
This looks good, but it seems to add a lot of unrelated whitespace changes. Are these necessary? I'm a bit confused whether this reformatting is on purpose.. it seems to add couple of spaces, but breaks consistency?

nkakouros replied:
I fixes some inconsistent alignment. I moved all comments a bit to the right to have the code breath. The count regex was 6 characters longer that the longest other regex and the most characters any other regex was separated from the comments was 5 characters. This meant I had to move them all to have them align with the new comment for the count regex.

liskin replied:
I think that's because `r'(limit:(?P<count>[0-9]+))?'` is longer than the rest and the comments wouldn't be aligned.

liskin replied:
(Yeah, it would be easier to review if the comments were moved in another commit, but I'm not sure that level of pedantry is warranted here. :-))

tbabej commented on taskwiki/viewport.py:
```diff
@@ -58,6 +67,8 @@ def __init__(self, line_number, cache, tw,
             constants.DEFAULT_SORT_ORDER
         )
 
+        self.count = count and int(count) or -1
```
The purpose of this boolean expression is not entirely clear to me. Maybe we could just have the default value of `count` set as None?

nkakouros replied:
I change the default to `None`. The boolean `and` is needed to convert the count to an integer. `int(None)` raises an error. Originally, I set the default to `-1` to be sure that the count is an integer.

tbabej commented on taskwiki/viewport.py:
```diff
@@ -29,7 +29,7 @@ class ViewPort(object):
     """
 
-    meta_tokens = ("-VISIBLE",)
+    meta_tokens = ('-VISIBLE',)
```
This commit can be rebased and squashed into the first one.

nkakouros replied:
I will squash (or you can via he github interface) everything into a single commit, no need to pollute the log.

nkakouros replied:
I just squashed everything into the first commit.

liskin commented on doc/taskwiki.txt:
```diff
@@ -382,8 +382,22 @@ viewport definition:
+~   == List only the 5 first tasks | project:Home limit:5 ==
+
+This will instruct taskwiki to limit the list of tasks to 5. The limit keyword
+will not be passed over to taskwarrior and is consumed only only by taskwiki.
```
```suggestion
will not be passed over to taskwarrior and is consumed only by taskwiki.
```

liskin commented on doc/taskwiki.txt:
```diff
@@ -382,8 +382,22 @@ viewport definition:
+You can limit the number of tasks listed by using the limit filter after the
+sorting identifier:
+
+~   == List only the 5 first tasks | project:Home limit:5 ==
```
This example doesn't really demonstrate that `limit:5` must come after `#X` or `$Y`, and makes it look like `limit:5` is just another taskwarrior filter (which is what I think @tbabej meant by “just "overriding" the behavior of the limit:<limit> clause that TW users are already familiar with”). I'm afraid this will confuse people.

nkakouros replied:
I clarified the docs. There is the issue that it looks a bit unnatural to have sth like:

```
# Tasks | project:mine other:filter +tag1 $T limit:3
```

I would expect taskwiki stuff to be last and the above seems like taskwarrior syntax is split.

nkakouros commented:
@tbabej It would be great if you could spare a couple of minutes to review the last changes!

liskin commented on doc/taskwiki.txt:
```diff
@@ -382,8 +382,28 @@ viewport definition:
+
+You can limit the number of tasks listed by using the limit filter:
+
+~   == List only the 5 first tasks | project:Home $T limit:5 ==
```
I think you only wanted to mention `$T` in the second example.

nkakouros replied:
Yes, you are right!

tbabej commented:
Merged. Thanks for the patience here, I'm juggling quite a bit of these open source projects :slightly_smiling_face:
